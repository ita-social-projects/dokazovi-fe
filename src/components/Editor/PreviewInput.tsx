import React, { useState } from 'react';
import { TextField } from '@material-ui/core';
import { useTranslation } from 'react-i18next';
import { useEffectExceptOnMount } from '../../old/lib/hooks/useEffectExceptOnMount';
import { MAX_PREVIEW_LENGTH } from '../../old/lib/constants/editors';
import { langTokens } from '../../locales/localizationInit';
import { useStyle } from '../../views/postCreation/RequiredFieldsStyle';

const truncText = (str: string) => {
  return str.slice(0, MAX_PREVIEW_LENGTH);
};

export interface IPreviewInputProps {
  initialPreview: string;
  editorTextContent?: string;
  initialWasManuallyChanged: boolean;
  onPreviewChange: (value: string) => void;
  onManuallyChanged?: () => void;
  disableAutoChanges?: () => void;
}

const PreviewInput: React.FC<IPreviewInputProps> = ({
  initialPreview,
  editorTextContent,
  initialWasManuallyChanged,
  onPreviewChange,
  onManuallyChanged,
  disableAutoChanges,
}) => {
  const { t } = useTranslation();

  const [textFieldValue, setTextFieldValue] = useState<string>(initialPreview);
  const [isTextFieldManuallyChanged, setIsTextFieldManuallyChanged] = useState(
    initialWasManuallyChanged,
  );
  const [isPreviewValid, setIsPreviewValid] = useState(
    textFieldValue.length <= MAX_PREVIEW_LENGTH,
  );

  useEffectExceptOnMount(() => {
    setIsPreviewValid(textFieldValue.length <= MAX_PREVIEW_LENGTH);
    onPreviewChange(truncText(textFieldValue));
  }, [textFieldValue]);

  const handleManualChange = (value: string) => {
    if (!isTextFieldManuallyChanged) {
      setIsTextFieldManuallyChanged(true);
      if (onManuallyChanged) onManuallyChanged();
    }
    if (disableAutoChanges) {
      disableAutoChanges();
    }
    setTextFieldValue(value);
  };

  const classes = useStyle();

  return (
    <TextField
      data-testid="preview-input"
      aria-label="minimum height"
      value={textFieldValue}
      multiline
      label={t(langTokens.editor.materialCardText)}
      variant="outlined"
      onChange={(e) => {
        handleManualChange(e.target.value);
      }}
      InputLabelProps={{
        classes: { formControl: classes.requiredAuthorField },
      }}
      InputProps={{
        inputProps: {
          style: {
            height: '165px',
          },
        },
      }}
      style={{
        width: '100%',
      }}
      error={!isPreviewValid}
      helperText={
        (!isPreviewValid &&
          `${t(langTokens.editor.maxTextlength, {
            count: MAX_PREVIEW_LENGTH,
          })}!`) ||
        t(langTokens.editor.textLength, { count: textFieldValue.length })
      }
    />
  );
};

export default React.memo(PreviewInput);
