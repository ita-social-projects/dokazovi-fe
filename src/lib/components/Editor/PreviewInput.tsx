import React, { useState } from 'react';
import { TextField } from '@material-ui/core';
import useEffectExceptOnMount from '../../hooks/useEffectExceptOnMount';
import { MAX_PREVIEW_LENGTH } from '../../constants/editors';

const truncText = (str: string) => {
  return str.slice(0, MAX_PREVIEW_LENGTH);
};

export interface IPreviewInputProps {
  initialPreview: string;
  editorTextContent?: string;
  initialWasManuallyChanged: boolean;
  onPreviewChange: (value: string) => void;
  onManuallyChanged?: () => void;
}

const PreviewInput: React.FC<IPreviewInputProps> = ({
  initialPreview,
  editorTextContent,
  initialWasManuallyChanged,
  onPreviewChange,
  onManuallyChanged,
}) => {
  const [textFieldValue, setTextFieldValue] = useState<string>(initialPreview);
  const [isTextFieldManuallyChanged, setIsTextFieldManuallyChanged] = useState(
    initialWasManuallyChanged,
  );
  const [isPreviewValid, setIsPreviewValid] = useState(
    textFieldValue.length <= MAX_PREVIEW_LENGTH,
  );

  useEffectExceptOnMount(() => {
    const setPreviewFromEditor = () => {
      if (!initialWasManuallyChanged && editorTextContent) {
        setTextFieldValue(truncText(editorTextContent));
      }
    };
    setPreviewFromEditor();
  }, [editorTextContent]);

  useEffectExceptOnMount(() => {
    setIsPreviewValid(textFieldValue.length <= MAX_PREVIEW_LENGTH);
    onPreviewChange(truncText(textFieldValue));
  }, [textFieldValue]);

  const handleManualChange = (value: string) => {
    if (!isTextFieldManuallyChanged) {
      setIsTextFieldManuallyChanged(true);
      if (onManuallyChanged) onManuallyChanged();
    }
    setTextFieldValue(value);
  };

  return (
    <TextField
      aria-label="minimum height"
      value={textFieldValue}
      multiline
      label="Текст картки матеріалу"
      variant="outlined"
      onChange={(e) => {
        handleManualChange(e.target.value);
      }}
      InputProps={{
        inputProps: {
          style: {
            height: '165px',
          },
        },
      }}
      style={{
        width: '100%',
      }}
      error={!isPreviewValid}
      helperText={
        (!isPreviewValid &&
          `Максимальна довжина тексту: ${MAX_PREVIEW_LENGTH} символів!`) ||
        `Довжина тексту: ${textFieldValue.length} символів`
      }
    />
  );
};

export default React.memo(PreviewInput);
