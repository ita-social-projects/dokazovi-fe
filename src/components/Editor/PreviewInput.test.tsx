import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import PreviewInput from './PreviewInput';

const mockOnPreviewChange = jest.fn();
const mockOnManuallyChanged = jest.fn();

test('component renders properly with only required props', () => {
  render(
    <PreviewInput
      initialPreview="test"
      initialWasManuallyChanged
      onPreviewChange={mockOnPreviewChange}
    />,
  );

  expect(screen.getByTestId('preview-input')).toBeInTheDocument();
  expect(screen.getByTestId('preview-input')).toMatchSnapshot();
});

test('component renders properly with all props', () => {
  render(
    <PreviewInput
      initialPreview="initial preview text"
      initialWasManuallyChanged={false}
      onPreviewChange={mockOnPreviewChange}
      editorTextContent="Editor text content"
      onManuallyChanged={mockOnManuallyChanged}
    />,
  );

  expect(screen.getByTestId('preview-input')).toBeInTheDocument();
  expect(screen.getByTestId('preview-input')).toMatchSnapshot();
});

test('component renders initial preview text', () => {
  render(
    <PreviewInput
      initialPreview="test"
      initialWasManuallyChanged
      onPreviewChange={mockOnPreviewChange}
    />,
  );

  expect(screen.getByText('test')).toBeInTheDocument();
});

test('component changes initial text after user input', () => {
  render(
    <PreviewInput
      initialPreview="initial text"
      initialWasManuallyChanged={false}
      onPreviewChange={mockOnPreviewChange}
      onManuallyChanged={mockOnManuallyChanged}
    />,
  );
  fireEvent.change(screen.getByText('initial text'), {
    target: { value: 'test' },
  });
  expect(screen.getByText('test')).toBeInTheDocument();
});

test('component displays text length properly', () => {
  render(
    <PreviewInput
      initialPreview="initial text"
      initialWasManuallyChanged={false}
      onPreviewChange={mockOnPreviewChange}
      onManuallyChanged={mockOnManuallyChanged}
    />,
  );
  expect(
    screen.getByText(
      'Мінімальна довжина 50 символів. Довжина тексту: 12 символів',
    ),
  ).toBeInTheDocument();
  fireEvent.change(screen.getByText('initial text'), {
    target: { value: 'test' },
  });
  expect(
    screen.getByText(
      'Мінімальна довжина 50 символів. Довжина тексту: 4 символи',
    ),
  ).toBeInTheDocument();
});

test('component calls callback after user input', () => {
  render(
    <PreviewInput
      initialPreview="initial text"
      initialWasManuallyChanged={false}
      onPreviewChange={mockOnPreviewChange}
      onManuallyChanged={mockOnManuallyChanged}
    />,
  );
  fireEvent.change(screen.getByText('initial text'), {
    target: { value: 'test' },
  });
  expect(mockOnPreviewChange).toBeCalled();
  expect(mockOnManuallyChanged).toBeCalled();
});
