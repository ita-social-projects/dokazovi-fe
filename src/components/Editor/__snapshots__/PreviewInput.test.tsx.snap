// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component renders properly with all props 1`] = `
<div
  aria-label="minimum height"
  class="MuiFormControl-root MuiTextField-root"
  data-testid="preview-input"
  style="width: 100%;"
>
  <label
    class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl makeStyles-requiredAuthorField-14 MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-filled"
    data-shrink="true"
  >
    Текст картки матеріалу
  </label>
  <div
    class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl MuiInputBase-multiline MuiOutlinedInput-multiline"
  >
    <textarea
      aria-invalid="false"
      class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline MuiOutlinedInput-inputMultiline"
      rows="1"
      style="height: 165px; overflow: hidden;"
    >
      initial preview text
    </textarea>
    <textarea
      aria-hidden="true"
      class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline MuiOutlinedInput-inputMultiline"
      readonly=""
      style="visibility: hidden; position: absolute; overflow: hidden; height: 165px; top: 0px; left: 0px; transform: translateZ(0); width: 100%;"
      tabindex="-1"
    />
    <fieldset
      aria-hidden="true"
      class="PrivateNotchedOutline-root-21 MuiOutlinedInput-notchedOutline"
    >
      <legend
        class="PrivateNotchedOutline-legendLabelled-23 PrivateNotchedOutline-legendNotched-24"
      >
        <span>
          Текст картки матеріалу
        </span>
      </legend>
    </fieldset>
  </div>
  <p
    class="MuiFormHelperText-root MuiFormHelperText-contained MuiFormHelperText-filled"
  >
    Мінімальна довжина 50 символів. Довжина тексту: 20 символів
  </p>
</div>
`;

exports[`component renders properly with only required props 1`] = `
<div
  aria-label="minimum height"
  class="MuiFormControl-root MuiTextField-root"
  data-testid="preview-input"
  style="width: 100%;"
>
  <label
    class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl makeStyles-requiredAuthorField-2 MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-filled"
    data-shrink="true"
  >
    Текст картки матеріалу
  </label>
  <div
    class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl MuiInputBase-multiline MuiOutlinedInput-multiline"
  >
    <textarea
      aria-invalid="false"
      class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline MuiOutlinedInput-inputMultiline"
      rows="1"
      style="height: 165px; overflow: hidden;"
    >
      test
    </textarea>
    <textarea
      aria-hidden="true"
      class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline MuiOutlinedInput-inputMultiline"
      readonly=""
      style="visibility: hidden; position: absolute; overflow: hidden; height: 165px; top: 0px; left: 0px; transform: translateZ(0); width: 100%;"
      tabindex="-1"
    />
    <fieldset
      aria-hidden="true"
      class="PrivateNotchedOutline-root-9 MuiOutlinedInput-notchedOutline"
    >
      <legend
        class="PrivateNotchedOutline-legendLabelled-11 PrivateNotchedOutline-legendNotched-12"
      >
        <span>
          Текст картки матеріалу
        </span>
      </legend>
    </fieldset>
  </div>
  <p
    class="MuiFormHelperText-root MuiFormHelperText-contained MuiFormHelperText-filled"
  >
    Мінімальна довжина 50 символів. Довжина тексту: 4 символи
  </p>
</div>
`;
